core/types/auxpow.go

```diff
@@ -0,0 +1,916 @@
+func (at *AuxTemplate) ProtoDecode(data *ProtoAuxTemplate) error {
+	if data == nil {
+		return nil
+	}
+
+	at.powID = PowID(data.GetChainId())
+
+	// Copy PrevHash (32 bytes)
+	if len(data.GetPrevHash()) == 32 {
+		copy(at.prevHash[:], data.GetPrevHash())
+	}
+
+	at.auxPow2 = data.GetAuxPow2()
+	at.version = data.GetVersion()
+	at.nBits = data.GetNbits()
+	at.nTimeMask = NTimeMask(data.GetNtimeMask())
+	at.height = data.GetHeight()
+	at.coinbaseOut = data.GetCoinbaseOut()
+
+	// Copy merkle branch
+	at.merkleBranch = make([][]byte, len(data.GetMerkleBranch()))
+	for i, hash := range data.GetMerkleBranch() {
+		at.merkleBranch[i] = make([]byte, len(hash))
+		copy(at.merkleBranch[i], hash)
+	}
+
+	// Decode signer envelopes
+	at.sigs = make([]byte, len(data.GetSigs()))
+	copy(at.sigs, data.GetSigs())
+
+	return nil
+}
+
+// Hash returns the SHA256 hash of the AuxTemplate with signature fields set to nil
+// This is the same hash used for signing and verification
+func (at *AuxTemplate) Hash() [32]byte {
+	// Create a copy of the template without signatures for message hash calculation
+	// We need to work with the protobuf representation to properly exclude signature fields
+	protoTemplate := at.ProtoEncode()
+	tempTemplate := proto.Clone(protoTemplate).(*ProtoAuxTemplate)
+	tempTemplate.Sigs = nil
+
+	// Since for chains other than ravencoin, version can be grinded for mining,
+	// need to mask them before signing
+	versionMask := uint32(0xE0000000)
+	switch at.powID {
+	case SHA_BTC, SHA_BCH:
+		tempTemplate.Version = proto.Uint32(*tempTemplate.Version & versionMask)
+	}
+
+	// Marshal the template without signatures to get the message hash
+	templateData, err := proto.Marshal(tempTemplate)
+	if err != nil {
+		// Return zero hash on error - this should be handled by the caller
+		return [32]byte{}
+	}
+
+	// Calculate and return the message hash
+	return sha256.Sum256(templateData)
+}
+
+// VerifySignature verifies the composite MuSig2 signature over this AuxTemplate.
+// It tries all possible 2-of-3 signer index combinations. If 'signature' is nil or empty,
+// it falls back to using the embedded at.sigs. Returns true on any valid combination.
+func (at *AuxTemplate) VerifySignature() bool {
+	// Check Signature presence
+	if len(at.sigs) == 0 {
+		log.Global.Warn("Sig is empty, signature failed")
+		return false
+	}
+
+	// Build the canonical message from the template (Sigs excluded internally)
+	msgHash := at.Hash()
+	message := msgHash[:]
+
+	// Try all 2-of-3 combinations in both orders (order-dependent)
+	combinations := [][]int{
+		{0, 1}, {1, 0},
+		{0, 2}, {2, 0},
+		{1, 2}, {2, 1},
+	}
+	for _, signerIndices := range combinations {
+		if err := musig2.VerifyCompositeSignature(message, at.sigs, signerIndices); err == nil {
+			return true
+		}
+	}
+
+	log.Global.Warn("Signature verification failed for all signer combinations", at)
+
+	return false
+}
+
+type AuxPowHeader struct {
```

Comment:
Out of curiosity, why did we wrap the AuxHeaderData in a AuxPowHeader struct?

Reply:
two reasons,
 1) if we want to have state for the auxheader for example cache or something, not possible with interface
 2) so that you can hide the implementations private and have a single public function operating on the struct, which technically we are not doing, i can make it private